From a68f43ffe7be63cb2dd5839b493547249211ad0a Mon Sep 17 00:00:00 2001
From: Hugues Fruchet <hugues.fruchet@foss.st.com>
Date: Mon, 18 Sep 2023 13:28:32 +0200
Subject: [PATCH 2/4] v4l2codecs: fix gst_v4l2_encoder_set_src_fmt() call

Set source format regarding state->vinfo and call it
right before setting sink format.
Fixes issues when encoding from YUY2 and I420 formats.

Signed-off-by: Hugues Fruchet <hugues.fruchet@foss.st.com>
Upstream-Status: Pending
---
 sys/v4l2codecs/gstv4l2codecvp8enc.c | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/sys/v4l2codecs/gstv4l2codecvp8enc.c b/sys/v4l2codecs/gstv4l2codecvp8enc.c
index 71c7dd2..2d476b3 100644
--- a/sys/v4l2codecs/gstv4l2codecvp8enc.c
+++ b/sys/v4l2codecs/gstv4l2codecvp8enc.c
@@ -285,6 +285,14 @@ gst_v4l2_codec_vp8_enc_set_format (GstVideoEncoder * encoder,
 
   gst_v4l2_codec_vp8_enc_reset_allocation (self);
 
+  if (!gst_v4l2_encoder_set_src_fmt (self->encoder, &state->info,
+          V4L2_PIX_FMT_VP8_FRAME)) {
+    GST_ELEMENT_ERROR (self, CORE, NEGOTIATION, ("Unsupported pixel format"),
+        ("No support for %ux%u format VP8", state->info.width,
+            state->info.height));
+    return FALSE;
+  }
+
   if (!gst_v4l2_encoder_select_sink_format (self->encoder, &state->info,
           &self->vinfo)) {
     GST_ELEMENT_ERROR (self, CORE, NEGOTIATION,
@@ -295,14 +303,6 @@ gst_v4l2_codec_vp8_enc_set_format (GstVideoEncoder * encoder,
     return FALSE;
   }
 
-  if (!gst_v4l2_encoder_set_src_fmt (self->encoder, &self->vinfo,
-          V4L2_PIX_FMT_VP8_FRAME)) {
-    GST_ELEMENT_ERROR (self, CORE, NEGOTIATION, ("Unsupported pixel format"),
-        ("No support for %ux%u format VP8", state->info.width,
-            state->info.height));
-    return FALSE;
-  }
-
   self->width = state->info.width;
   self->height = state->info.height;
   gst_v4l2_codec_vp8_enc_buffers_allocation (encoder);
-- 
2.25.1

